### ISSUE TITLE
Integration with custom AWS cognito/ amplify

### ISSUE THREAD

**Issue by hamidre13** [nextauthjs/next-auth#2525]
### Description 📓

We love next-auth and also AWS Cognito, but the hosted UI for AWS Cognito is ugly. As of now we could not find an easy way to have our own custom UI for AWS Cognito that can also integrates with next-auth after login. That would be great to have API to support that. 

### How to reproduce ☕️

N/A

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

**Comment by balazsorban44** [comment]
Could you elaborate more on what you would expect? That is a very vague description, and I don't have any experience with Cognito myself.

You can create your own signIn page if that is helpful, just use a button with `signIn("cognito")` and that will log you directly into Cognito.

**Comment by hamidre13** [comment]
For sure. 
you can use Cognito in two ways. 

1. Hosted login pages
2. use your own custom UI with the help of amazon-cognito-identity-js or aws-amplify package

With next Auth and  `signIn("cognito")` . you will be redirected to an ugly plage like this:
<img width="1412" alt="IMG_7763" src="https://user-images.githubusercontent.com/2216126/129411851-9a30796d-7335-4564-9166-6484e7a923a8.png">

This page is the hosted login page for AWS Cognito and has very limited customization capability. This solution will not work if we want to host our own login/logout page and have control over all the design.

Now if we want to have control over the design we will have to use the packages that I mentioned above. Now after login we like to give the session info to the next auth to handle the rest for us instead of us taking care of session management, URL management,... 

Basically we can use those package to login but we want to handle to token and user info to next-auth after that

**Comment by hamidre13** [comment]
AWS lets you either use the hosted sing up page or use an npm package (amplify, aws-conito) to the signup/sign in. We can definitely design the signup/sing in page but we like to then hand over our access token and refresh token to next-auth. Thus , what we are looking for is not and actual page design but an API in back end to tell next-auth that the user is signed in with following access, and refresh tokens .

**Comment by chan-vince** [comment]
> AWS lets you either use the hosted sing up page or use an npm package (amplify, aws-conito) to the signup/sign in. We can definitely design the signup/sing in page but we like to then hand over our access token and refresh token to next-auth. Thus , what we are looking for is not and actual page design but an API in back end to tell next-auth that the user is signed in with following access, and refresh tokens .

Did you end up getting anywhere with solving this? I'm looking for something that's exactly what you describe 🙂

**Comment by nk9** [comment]
I obviously agree. However, I will say that since I didn't have this option at the time I needed it, I did drop NextAuth entirely and switch to Amplify UI. After some stumbling around, I got it working quite well.

While I'd prefer NextAuth if all I needed was a simple sign in/sign up/logout flow, I needed Cognito's support for MFA using TOTP tokens. I don't believe NextAuth would have supported that, at least not out of the box, whereas Amplify UI's [Authenticator component](https://ui.docs.amplify.aws/components/authenticator) supports this flow automatically. It even includes a QR code to set up the TOTP app (e.g. Authy).

So I'd just urge people to fully understand their requirements before picking a library to interact with their IdP.

**Comment by yp717** [comment]
Yeah that makes sense 😊 you can use some amplify features to interface with cognito from the core and auth modules. Do you know why you need to use the amplify cli setup for the UI components but not to use the auth stand alone?

**Comment by nk9** [comment]
> Do you know why you need to use the amplify cli setup for the UI components but not to use the auth stand alone?

Not sure what you mean, exactly. I am using Amplify UI without the CLI.

**Comment by joseph-mccombs** [comment]
I also agree that this should be opened... I also am trying to add a custom template due to Cognito's clunky UI, so that we can support a custom flow involving inviteCodes. All we can currently use is the template provided by Cognito. @nk9 do you have an example of using amplify with cognito you could point me to?

**Comment by nk9** [comment]
Yep, [on Stack Overflow](https://stackoverflow.com/a/71125057/1749551).

**Comment by zul-yang** [comment]
@nk9 , apologies for reviving an old issue. When using Amplify UI, are you limited to storing the tokens in localstorage as well? I remember being this an issue: https://github.com/aws-amplify/amplify-js/issues/3224

**Comment by nk9** [comment]
I am not sure, I'm just using Amplify's `Authenticator` component. It's taking care of any cookie management and storage itself, and I just fetch the tokens when needed [using the API](https://ui.docs.amplify.aws/react/connected-components/authenticator/headless#access-authenticated-user).

**Comment by hdntc** [comment]
This not being possible is a bit limiting for the use cases of this library w/ cognito

**Comment by zarif-al** [comment]
Hi!. We also came across this issue while trying to use next-auth with aws-cognito. 

Currently, instead of using `CognitoProvider` we are using the `CredentialsProvider` paired with `aws-amplify`. I am not sure if this is the best way, but its working for us so far. We can have a custom sign-in experience while using cognito.

**Comment by Lagyu** [comment]
This can be another sample implementation with `CredentialsProvider`.
They use `CognitoIdentityProviderClient` to authorize users.

https://github.com/cds-snc/platform-forms-client/blob/develop/pages/api/auth/%5B...nextauth%5D.ts

**Comment by Its-suLav-D** [comment]
> AWS lets you either use the hosted sing up page or use an npm package (amplify, aws-conito) to the signup/sign in. We can definitely design the signup/sing in page but we like to then hand over our access token and refresh token to next-auth. Thus , what we are looking for is not and actual page design but an API in back end to tell next-auth that the user is signed in with following access, and refresh tokens .

Did we make any progress on this ?

**Comment by razzededge** [comment]
I would also love to have this - but it seems that for now I need to implement a custom mix of CredentialsProvider with Amplify... which is a suboptimal solution

**Comment by maorhoox** [comment]
More than two years later and still no integrated solution from the Vercel team?

**Comment by TechLipefi** [comment]
I also have this issue especially hard to use Cognito Provider and Social Providers inside Cognito with custom UI

> More than two years later and still no integrated solution from the Vercel team?

I suppose from business perspective it is not in focus just because Cognito is a plain competitor)

**Comment by maniraj1234** [comment]
More than 2 years, still no activity :( and this is common major problem for everyone who uses cognito.

**Comment by razzededge** [comment]
> More than 2 years, still no activity :( and this is common major problem for everyone who uses cognito.

This is why I opted for database backend - well I've used DynamoDB Adapter.
Cognito has some major flaws when you want to use social logins - it's almost impossible to merge accounts when somebody created their account using social login - other way around is possible, but you will end up with duplicated accounts based on that. Think about this before you use cognito in new project.

**Comment by balazsorban44** [comment]
> More than two years later and still no integrated solution from the Vercel team?

Don't really understand this sentiment. 🙂 I work at Vercel, but this project has no other affiliation other than they let me work on it. All decisions are made by the Auth.js team.

There is no "behind-the-scene" dark plan or competition. Already answered here https://github.com/nextauthjs/next-auth/issues/2525#issuecomment-899569064

NextAuth.js is not the correct place for creating your custom Cognito UI. It's a different abstraction. It's cool that they give you SDKs to build your own though!

If someone builds it, we can probably promote it in our docs.

**Comment by jamaybyrone** [comment]
@balazsorban44 like [this ]( https://github.com/jamaybyrone/Next-Auth-Custom-Cognito)?
